File: frontend/src/stores/performance.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import api from "../api/client";

export interface PerformanceSummary {
  total_requests: number;
  rpm: number;
  tpm: number;
  avg_latency_ms: number;
  p50_ms: number;
  p95_ms: number;
  p99_ms: number;
  avg_ttft_ms: number | null;
  ttft_count: number;
  error_rate: number;
  cache_hit_rate: number;
}

export interface PerformanceBucket {
  bucket: string;
  request_count: number;
  rpm: number;
  tpm: number;
  avg_latency_ms: number;
  avg_ttft_ms: number | null;
  error_count: number;
}

export interface PerformanceBreakdown {
  provider_id: number | null;
  provider_name: string;
  model: string;
  requests: number;
  tokens: number;
  avg_latency_ms: number;
  cost: number;
}

export interface PerformanceData {
  summary: PerformanceSummary;
  timeseries: PerformanceBucket[];
  by_provider: PerformanceBreakdown[];
  by_model: PerformanceBreakdown[];
  top_models_by_cost: PerformanceBreakdown[];
}

export const usePerformanceStore = defineStore("performance", () => {
  const data = ref<PerformanceData | null>(null);
  const loading = ref(false);
  const error = ref<string | null>(null);

  async function fetchPerformance(params?: Record<string, any>) {
    loading.value = true;
    error.value = null;
    try {
      const { data: res } = await api.get("/performance", { params });
      data.value = res as PerformanceData;
    } catch (err: any) {
      error.value = err?.response?.data?.error || err?.message || "Failed to load performance";
    } finally {
      loading.value = false;
    }
  }

  function clearError() {
    error.value = null;
  }

  return { data, loading, error, fetchPerformance, clearError };
});

File: backend/internal/models/performance.go
package models

type PerformanceQueryParams struct {
	ProviderID  *int64 `form:"provider_id"`
	From        string `form:"from"`
	To          string `form:"to"`
	Granularity string `form:"granularity"`
}

type PerformanceSummary struct {
	TotalRequests int64    `json:"total_requests"`
	RPM           float64  `json:"rpm"`
	TPM           float64  `json:"tpm"`
	AvgLatencyMs  float64  `json:"avg_latency_ms"`
	P50Ms         float64  `json:"p50_ms"`
	P95Ms         float64  `json:"p95_ms"`
	P99Ms         float64  `json:"p99_ms"`
	AvgTTFTMs     *float64 `json:"avg_ttft_ms"`
	TTFTCount     int64    `json:"ttft_count"`
	ErrorRate     float64  `json:"error_rate"`
	CacheHitRate  float64  `json:"cache_hit_rate"`
}

type PerformanceBucket struct {
	Bucket       string   `json:"bucket"`
	RequestCount int64    `json:"request_count"`
	RPM          float64  `json:"rpm"`
	TPM          float64  `json:"tpm"`
	AvgLatencyMs float64  `json:"avg_latency_ms"`
	AvgTTFTMs    *float64 `json:"avg_ttft_ms"`
	ErrorCount   int64    `json:"error_count"`
}

type PerformanceBreakdown struct {
	ProviderID   *int64  `json:"provider_id"`
	ProviderName string  `json:"provider_name"`
	Model        string  `json:"model"`
	Requests     int64   `json:"requests"`
	Tokens       int64   `json:"tokens"`
	AvgLatencyMs float64 `json:"avg_latency_ms"`
	Cost         float64 `json:"cost"`
}

type PerformanceResponse struct {
	Summary         PerformanceSummary     `json:"summary"`
	Timeseries      []PerformanceBucket    `json:"timeseries"`
	ByProvider      []PerformanceBreakdown `json:"by_provider"`
	ByModel         []PerformanceBreakdown `json:"by_model"`
	TopModelsByCost []PerformanceBreakdown `json:"top_models_by_cost"`
}
